src/tof/source.py

```diff
@@ -75,37 +67,24 @@ def _make_pulses(
     t_dim = "birth_time"
     w_dim = "wavelength"
 
-    p_time = _convert_coord(p_time, unit=TIME_UNIT, coord=t_dim)
-    p_wav = _convert_coord(p_wav, unit=WAV_UNIT, coord=w_dim)
-    sampling = int(sampling)
+    if p is None:
+        if None in (p_time, p_wav):
+            raise ValueError(
+                "Either p (2D) or both p_time (1D) and p_wav (1D) must be supplied."
+            )
+        p = (p_wav / p_wav.data.sum()) * (p_time / p_time.data.sum())
+    else:
+        p = p.copy(deep=False)
 
-    tmin = p_time.coords[t_dim].min()
-    tmax = p_time.coords[t_dim].max()
+    p.coords[t_dim] = p.coords[t_dim].to(dtype=float, unit=TIME_UNIT)
+    p.coords[w_dim] = p.coords[w_dim].to(dtype=float, unit=WAV_UNIT)
+
+    tmin = p.coords[t_dim].min()
+    tmax = p.coords[t_dim].max()
     if wmin is None:
-        wmin = p_wav.coords[w_dim].min()
+        wmin = p.coords[w_dim].min()
     if wmax is None:
-        wmax = p_wav.coords[w_dim].max()
-
-    time_interpolator = interp1d(p_time, dim=t_dim, fill_value="extrapolate")
```

Comment:
Removed the internal linear interpolation and the somewhat unusual `sampling` arg.
We now require users to provide a distribution with enough data points.